### PR TITLE
Replace Django/MySQL with Flask/SQLite

### PR DESCRIPTION
**Pull request by kbumsik**
After a long thought I feel I need to change many things. I switched Django to Flask for web framework. Also I decided to use SQLite rather than MySQL. I also started using `python3`.

So far. I implemented basic login/logout. Take a look. By the way. Now program can run on Windows and MacOS very easily. I replaced all shell scripts with python. All you need to install is SQLite cli tools to run it on your macbook. 
### Django

Django itself is a great MVC framework to build a website quickly.
However, Django is not suitable for our project for some reasons:
1. Pattern design mismatch: Django force developer to follow their own MVC pattern. However, I will write SQL statements directly for `Model` and @cjackie will use ReactJS through API for `View`. So we do not use `Model` and `View` of Django. So we do not take any advantages from Django.
2. Too much "magical" things to learn: There are so many thing to use, and its documentation is not useful enough. Eventually the code written for Django does not look "pythonic".
3. Not for API-based web development: Django's routing seems a bit odd.
### MySQL
1. Non-standard SQL: MySQL does not follow SQL standard (such as SQL92) at all. Historically MySQL focused on performance than standard conformance. SQL statements gets different with MySQL version. (we already had that problem).
2. Stored Procedure approach is too static: This is my problem though. Using stored procedures is too static to deal with JSON. I failed to combine SPs, transaction, and JSON altogether.
### Flask

Flask is microframework: it does not use `View` and `Model`. So design get more flexible.
1. Easy and lightweight: Look at `runserver.py`, `wolfie_home/webpages.py`, and `wolfie_home/apt_web.py` I am pretty much sure there is nothing you cannot understand. 
2. Routing: Flask uses decorators for routing like this:

```
@webpage.route('/index')
def main():
    return "Hello world"
```

I like this way. This way also used in `.NET MVC framwork` too, [using attributes (similar to decorator in python)](https://blogs.msdn.microsoft.com/webdev/2013/10/17/attribute-routing-in-asp-net-mvc-5/). Django separates routing objects and function, making harder to track. Also Flask has more flexible routing. This is better for API-base design.
### SQLite

SQLite is another famous RDBMS, but it is significantly different from ordinal RDBMS: it uses a single file for database (database/wolfie_home.db). 
1. Intended to used in embedded system/single application: It is very lightweight: the program is less than 350KiB only. The db is a single file so each application can run SQLite. This is used as internal DB engine in many games, application like firfox. SQLite also default database engine in iOS and Android. It can replace fopen() ( @cjackie, you probably want to use it in your android app). Also DB is just a single file, so you can easliy move and delete.
2. `sqlite3` is one of python's standard library: We don't need to install anything. just `import sqlite3`. We do not need to worry about any dependencies. But you might want to `sudo apt-get install sqlite3` for sqlite3's command-line tools.

SQLite does not have stored function/stored procedures. So I need to write SQL statements in python. (See `database/service`). It looks like I am executing SQL as string, but don't worry- it is quite SQL-injection safe. (see http://stackoverflow.com/a/29528839/4661625).

### CODE REVIEW

**Comment by cjackie** [comment]
😄 ,I will review it carefully by the end of this weekend. I think It looks great!!!

**Comment by kbumsik** [comment]
For front-end, the JSON response always include "msg" or "errmsg" entry, depending on response state. Front-end developers always can get messages from server. Refer `<div id="msg">` and `<div id="errmsg">`  in `templates`, `user.js`, and `wolfie_home/common.py`.

**Comment by cjackie** [comment]
@yiochen , you might like to take a look at this too. I know you have done SQLite before.

**Comment by cjackie** [comment]
> Pattern design mismatch: Django force developer to follow their own MVC pattern. However, I will write SQL statements directly for Model and @cjackie will use ReactJS through API for View. So we do not use Model and View of Django. So we do not take any advantages from Django.

Yes, I agree that we will not use Model and View in Django. these feature and template feature in Django, we will not use much. 

> Too much "magical" things to learn: There are so many thing to use, and its documentation is not useful enough. Eventually the code written for Django does not look "pythonic".

Yes, a lot of magical things. BUT, we do not need to learn these magic things. We should think of it as a black box(but you can kind of know how it is implemented), just know how to use it. Is it the one of main advantages of object oriented programming? Working with existing codebase, usually we had to work with black box a lot time.

> Not for API-based web development: Django's routing seems a bit odd.

Yes, it is awkward. But we can work with it.

Overall, we can switch to Flask. In the future, you might appreciate Django.

# 

> Stored Procedure approach is too static: This is my problem though. Using stored procedures is too static to deal with JSON. I failed to combine SPs, transaction, and JSON altogether.

I have looked at your new way of interacting with database, it seems like we are using ad-hoc approach again.   😄   #2 Yes, it is definitely easier to do data manipulation in high level language, than doing it in sql language.

# 

In summary, Since I will be doing front-end mainly, back-end is okay to me, as long as it provides a set of API I can interact with. Flask and SQLite is easier for you to do, we should use it. To me, it looks straight-forward too. 😄

**Comment by kbumsik** [comment]
Well you know, Maybe I didn't like non-native things like ORM so I might be just trying to ditch everything about Django to be honest lol. Speaking of ORM, I know can be run any RDBMS so we don't need change the SQL code when switching it. But in other words, ORM is bounded with a framework so we need to study whole different ORM when we switch it. At least SQL has standards.

The main advantage of SP is that it is "stored" in a database and then the SQL statements in SP is compiled to be executed very quickly. By using Raw SQLs (Ad-hoc) or ORM we have a slight disadvantage in speed.
More importantly. The major concern is SQL-Injection. Ad-hoc approach is very vulnerable to SQL injection attack. For example:

```
string = "SELECT * FROM Users WHERE name = %s;" %name_to_find
cur.execute(string) 
```

if `name_to_find` is '"John"; DROP TABLE Data'
the raw SQL become this:

`cur.execute("SELECT * FROM Users WHERE name = "John"; DROP TABLE Data;")`

And then we lost all records in `Data` table... In case of SP, the whole statement is already compiled to it is relatively safe from that attack. ORM is also quite safe cuz it does not use SQL statement directly. 

However, I found that Python's DB-API 2.0 ([PEP249](https://www.python.org/dev/peps/pep-0249/)standard python interface to work with DB, so either MySQL and SQLite has the same API interface by adopting it) Offers a safer way to do that. Instead of the above code, we use placeholder like `?`:

```
cur.execute( "SELECT * FROM Users WHERE name = ?", (name_to_find, )) 
```

By the [documentation](https://docs.python.org/3.5/library/sqlite3.html) this way will be safe. So we don't need to worry about SQL-level security much. Anyway, it is kinda funny to go back to Ad-hoc lol

By the way, have you try running my code? Is that working well in your machine too?

**Comment by cjackie** [comment]
> The major concern is SQL-Injection. Ad-hoc approach is very vulnerable to SQL injection attack

Yes, I'm aware of this kind of attack.

> By the way, have you try running my code? Is that working well in your machine too?

Yes, I did. It worked. I was able to login with dummy user account! There is a small typo in README file, I fixed it.

**Comment by kbumsik** [comment]
Oh thanks for the fix :)
